lib/types/db.ts

```diff
@@ -0,0 +1,53 @@
+import Sequelize, { DataTypeAbstract, DefineAttributeColumnOptions } from 'sequelize';
+import { db } from './index';
+
+export type SequelizeAttributes<T extends { [key: string]: any }> = {
+  [P in keyof T]: string | DataTypeAbstract | DefineAttributeColumnOptions
+};
+
+export type CurrencyFactory = {
```

Comment:
I always think of a "factory" in code as something that creates other objects. Here it seems to be used to describe a base set of properties that are used to create the complete version of an object, for example here the `CurrencyInstance`. I think maybe replacing "Factory" with "Request" (or a similar term) for the naming conventions in this file would make more sense - as in a "request" to create an order must have the following properties whereas an "instance" of an order must additionally have some other properties. Same would go for `Peer`, `Order`, and `Pair`.

Reply:
Factory - creating a new record
Attributes - the record
Instance - the record + operational properties

The usage of these “Factory” types is indeed for functions which creates objects (in the context of a database). Their usage in this file as a base set for “Attribute” and “Instance” is not very important - I did it only to avoid repeating property names. What’s important is that in the project level, they have a distinct role, and that's to the define the type for creating new records.

“Factory” might not be a perfect name, but I think that “Request” makes less sense. If you have other names in mind, tell me.

Sequelize actually requires the “Attribute” type when you want to create a new record. The only reason I used “Factory” in addition to that is because “Attribute” cannot support nullables, since it is being used also to define the table columns. 
So I couldn’t find a better way of supporting a more narrow object for creation, besides forking `@types/sequelize`, or adding this “Factory” type.

Reply:
Thanks for the explanation, makes sense now. I think it'd be helpful to summarize these in some tsdoc comments.

As for naming, I agree "Request" doesn't make that much sense, but "Factory" still throws me off. I looked for a while and didn't find any word that really perfectly fits. Still, some options to consider: "Base," "Core," or "Source." I think "Core" is my favorite, but pick any of these and we'll go with it. Or if you don't like any of them, just stick with "Factory" and make sure to comment the type to clarify what it's for.

Reply:
"Factory" still makes the most sense to me, being used in methods which create new records, so i'll stick to it for now.
I'll add some explanation in the beginning of this file.